File: paleo_workbench/ui/pages/map_layer_tree.py
from __future__ import annotations

from typing import Any

from PySide6.QtCore import Qt, Signal
from PySide6.QtWidgets import QFrame, QLabel, QTreeWidget, QTreeWidgetItem, QVBoxLayout

from paleo_workbench.ui import tokens
from paleo_workbench.viz.mapping_helpers import field_value

LAYER_KEYS = ("facies", "well", "line", "label")
LAYER_LABELS = {
    "facies": "相带",
    "well": "井",
    "line": "线",
    "label": "注记",
}


class MapLayerTree(QFrame):
    """Document list plus per-layer visibility and lock controls."""

    document_selected = Signal(object)
    layer_visibility_changed = Signal(str, bool)
    layer_lock_changed = Signal(str, bool)

    def __init__(self, parent=None):
        super().__init__(parent)
        self.setObjectName("MapLayerTree")
        self.setFixedWidth(240)

        layout = QVBoxLayout(self)
        layout.setContentsMargins(
            tokens.PANEL_PADDING,
            tokens.PANEL_PADDING,
            tokens.PANEL_PADDING,
            tokens.PANEL_PADDING,
        )
        layout.setSpacing(tokens.SPACE_2)

        title = QLabel("图件与图层")
        title.setObjectName("MapDockTitle")
        layout.addWidget(title)

        self.tree = QTreeWidget()
        self.tree.setObjectName("MapLayerTreeWidget")
        self.tree.setHeaderHidden(True)
        self.tree.setColumnCount(2)
        self.tree.setRootIsDecorated(True)
        layout.addWidget(self.tree, 1)

        self._documents: list[Any] = []
        self._active_document: Any = None
        self._doc_items: list[QTreeWidgetItem] = []
        self._layer_items: dict[str, QTreeWidgetItem] = {}
        self._layer_locked: dict[str, bool] = {k: False for k in LAYER_KEYS}
        self._layer_visible: dict[str, bool] = {k: True for k in LAYER_KEYS}
        self._suppress_item_changed = False

        self.tree.currentItemChanged.connect(self._on_current_item_changed)
        self.tree.itemChanged.connect(self._on_item_changed)

        self._rebuild_tree()

    def set_documents(self, documents: list | tuple | None) -> None:
        self._documents = list(documents or [])
        self._rebuild_tree()

    def set_active_document(self, document) -> None:
        self._active_document = document
        self._rebuild_tree()
        if document is None:
            return
        for item in self._doc_items:
            if item.data(0, Qt.ItemDataRole.UserRole) is document:
                self.tree.setCurrentItem(item)
                item.setExpanded(True)
                break

    def set_layer_locked(self, layer_key: str, locked: bool) -> None:
        if layer_key not in LAYER_KEYS:
            return
        if self._layer_locked.get(layer_key) == locked:
            return
        self._layer_locked[layer_key] = locked
        item = self._layer_items.get(layer_key)
        if item is not None:
            self._suppress_item_changed = True
            item.setText(1, "🔒" if locked else "")
            self._suppress_item_changed = False
        self.layer_lock_changed.emit(layer_key, locked)

    def layer_is_visible(self, layer_key: str) -> bool:
        return self._layer_visible.get(layer_key, True)

    def layer_is_locked(self, layer_key: str) -> bool:
        return self._layer_locked.get(layer_key, False)

    def _rebuild_tree(self) -> None:
        self._suppress_item_changed = True
        self.tree.clear()
        self._doc_items = []
        self._layer_items = {}

        root = QTreeWidgetItem(["图件"])
        root.setFlags(root.flags() & ~Qt.ItemFlag.ItemIsSelectable)
        self.tree.addTopLevelItem(root)
        root.setExpanded(True)

        active = self._active_document
        for doc in self._documents:
            name = field_value(doc, "name", "") or "未命名图件"
            item = QTreeWidgetItem([name])
            item.setData(0, Qt.ItemDataRole.UserRole, doc)
            item.setFlags(
                Qt.ItemFlag.ItemIsEnabled
                | Qt.ItemFlag.ItemIsSelectable
            )
            root.addChild(item)
            self._doc_items.append(item)

            if active is not None and doc is active:
                self._populate_layers(item)
                self._populate_reference_layers(item, doc)

        if active is None and self._doc_items:
            # Fallback: attach layers to last document if no explicit active
            self._populate_layers(self._doc_items[-1])
            self._populate_reference_layers(self._doc_items[-1], self._documents[-1])

        self._suppress_item_changed = False

    def _populate_layers(self, parent: QTreeWidgetItem) -> None:
        for key in LAYER_KEYS:
            label = LAYER_LABELS[key]
            layer_item = QTreeWidgetItem([label, ""])
            layer_item.setData(0, Qt.ItemDataRole.UserRole, ("layer", key))
            layer_item.setFlags(
                Qt.ItemFlag.ItemIsEnabled
                | Qt.ItemFlag.ItemIsUserCheckable
                | Qt.ItemFlag.ItemIsSelectable
            )
            visible = self._layer_visible.get(key, True)
            layer_item.setCheckState(
                0,
                Qt.CheckState.Checked if visible else Qt.CheckState.Unchecked,
            )
            if self._layer_locked.get(key, False):
                layer_item.setText(1, "🔒")
            parent.addChild(layer_item)
            self._layer_items[key] = layer_item
            parent.setExpanded(True)

    def _populate_reference_layers(self, parent: QTreeWidgetItem, document: Any) -> None:
        ref_layers = getattr(document, "reference_layers", [])
        if not ref_layers:
            return

        ref_group = QTreeWidgetItem(["参考图层"])
        ref_group.setFlags(Qt.ItemFlag.ItemIsEnabled)
        parent.addChild(ref_group)

        for layer in ref_layers:
            name = getattr(layer, "name", "未命名参考图层")
            status = getattr(layer, "status", "")
            
            if status == "offline":
                name = f"{name} (离线)"
            elif status == "failed":
                name = f"{name} (失败)"

            item = QTreeWidgetItem([name, ""])
            item.setData(0, Qt.ItemDataRole.UserRole, ("reference_layer", layer))
            item.setFlags(Qt.ItemFlag.ItemIsEnabled | Qt.ItemFlag.ItemIsSelectable)
            ref_group.addChild(item)
            
        ref_group.setExpanded(True)

    def _on_current_item_changed(
        self,
        current: QTreeWidgetItem | None,
        _previous: QTreeWidgetItem | None,
    ) -> None:
        if current is None:
            return
        data = current.data(0, Qt.ItemDataRole.UserRole)
        if data is None:
            return
        if isinstance(data, tuple) and data and data[0] == "layer":
            # Selecting a layer does not change document
            return
        if isinstance(data, tuple) and data and data[0] == "reference_layer":
            # Reference-layer rows are not map documents; treating them as one
            # emitted a raw tuple as the active document and crashed the host
            # page's load_document() with AttributeError.
            return
        # Document node
        if data is not self._active_document:
            self._active_document = data
            self._rebuild_tree()
            for item in self._doc_items:
                if item.data(0, Qt.ItemDataRole.UserRole) is data:
                    self.tree.blockSignals(True)
                    self.tree.setCurrentItem(item)
                    item.setExpanded(True)
                    self.tree.blockSignals(False)
                    break
            self.document_selected.emit(data)

    def _on_item_changed(self, item: QTreeWidgetItem, column: int) -> None:
        if self._suppress_item_changed:
            return
        data = item.data(0, Qt.ItemDataRole.UserRole)
        if not (isinstance(data, tuple) and data and data[0] == "layer"):
            return
        layer_key = data[1]
        if column == 0:
            visible = item.checkState(0) == Qt.CheckState.Checked
            if self._layer_visible.get(layer_key) != visible:
                self._layer_visible[layer_key] = visible
                self.layer_visibility_changed.emit(layer_key, visible)

File: libs/ui_map/src/map_layer_tree.cpp
#include <pwb/ui_map/map_layer_tree.hpp>

#include <QLabel>
#include <QTreeWidget>
#include <QVBoxLayout>

#include <pwb/ui_map/item_reconcile.hpp>

namespace pwb::ui_map {

namespace {

// Roles carried on items (Python packs ("layer", key) / ("reference_layer",
// obj) tuples into Qt.UserRole; C++ stores the tag + key in two roles).
constexpr int kTagRole = Qt::UserRole;        // "doc"|"layer"|"reference"
constexpr int kKeyRole = Qt::UserRole + 1;    // layer key or document key
constexpr int kDocIndexRole = Qt::UserRole + 2;  // index into documents_

QString qstr(const std::string& s) {
    return QString::fromUtf8(s.data(), static_cast<qsizetype>(s.size()));
}

}  // namespace

MapLayerTree::MapLayerTree(QWidget* parent) : QFrame(parent) {
    setObjectName(QStringLiteral("MapLayerTree"));
    setMinimumWidth(220);

    auto* layout = new QVBoxLayout(this);
    layout->setContentsMargins(12, 12, 12, 12);  // PANEL_PADDING
    layout->setSpacing(8);                       // SPACE_2

    auto* title = new QLabel(QStringLiteral("图件与图层"), this);
    title->setObjectName(QStringLiteral("MapDockTitle"));
    layout->addWidget(title);

    tree_ = new QTreeWidget(this);
    tree_->setObjectName(QStringLiteral("MapLayerTreeWidget"));
    tree_->setHeaderHidden(true);
    tree_->setColumnCount(2);
    tree_->setRootIsDecorated(true);
    layout->addWidget(tree_, 1);

    for (const char* key : kLayerKeys) {
        layer_locked_[key] = false;
        layer_visible_[key] = true;
    }

    connect(tree_, &QTreeWidget::currentItemChanged, this,
            [this](QTreeWidgetItem* current, QTreeWidgetItem*) {
                on_current_item_changed(current);
            });
    connect(tree_, &QTreeWidget::itemChanged, this,
            [this](QTreeWidgetItem* item, int column) {
                on_item_changed(item, column);
            });

    rebuild_tree();
}

void MapLayerTree::set_documents(const std::vector<Json>& documents) {
    documents_ = documents;
    rebuild_tree();
}

void MapLayerTree::set_active_document(const Json* document) {
    active_document_ = document;
    rebuild_tree();
    if (document == nullptr) {
        return;
    }
    // Select + expand the matching document row (key-stable lookup —
    // identity tags resolve through the populated documents_ vector).
    const std::string wanted_key = document_key(
        *document, identity_of(*document));
    for (QTreeWidgetItem* item : doc_items_) {
        if (item->data(0, kKeyRole).toString().toStdString() ==
            wanted_key) {
            tree_->setCurrentItem(item);
            item->setExpanded(true);
            break;
        }
    }
}

void MapLayerTree::set_layer_locked(const std::string& layer_key,
                                    bool locked) {
    const auto it = layer_locked_.find(layer_key);
    if (it == layer_locked_.end() || it->second == locked) {
        return;
    }
    it->second = locked;
    const auto item_it = layer_items_.find(layer_key);
    if (item_it != layer_items_.end() && item_it->second != nullptr) {
        suppress_item_changed_ = true;
        item_it->second->setText(
            1, locked ? QStringLiteral("⊘") : QString());
        suppress_item_changed_ = false;
    }
    emit layer_lock_changed(qstr(layer_key), locked);
}

bool MapLayerTree::layer_is_visible(const std::string& layer_key) const {
    const auto it = layer_visible_.find(layer_key);
    return it == layer_visible_.end() || it->second;
}

bool MapLayerTree::layer_is_locked(const std::string& layer_key) const {
    const auto it = layer_locked_.find(layer_key);
    return it != layer_locked_.end() && it->second;
}

const void* MapLayerTree::identity_of(const Json& document) const {
    // The identity tag is the document's index inside documents_ (stable
    // across reconciles of the same set; position-keyed like Python id()
    // for a persisted object).
    for (std::size_t i = 0; i < documents_.size(); ++i) {
        if (documents_[i] == document) {
            return reinterpret_cast<const void*>(i + 1);
        }
    }
    return nullptr;
}

void MapLayerTree::rebuild_tree() {
    suppress_item_changed_ = true;
    const bool was_blocked = tree_->signalsBlocked();
    tree_->blockSignals(true);
    const auto items = reconcile_tree_items(
        tree_, {kRootKey},
        [](const std::string&) {
            auto* root = new QTreeWidgetItem(
                QStringList{QStringLiteral("图件")});
            root->setFlags(root->flags() & ~Qt::ItemIsSelectable);
            return root;
        },
        [](QTreeWidgetItem* item, const std::string&) {
            if (item->text(0) != QStringLiteral("图件")) {
                item->setText(0, QStringLiteral("图件"));
            }
            item->setExpanded(true);
        });
    QTreeWidgetItem* root = items.at(kRootKey);

    std::vector<std::string> doc_keys;
    doc_keys.reserve(documents_.size());
    for (std::size_t i = 0; i < documents_.size(); ++i) {
        doc_keys.push_back(document_key(
            documents_[i], reinterpret_cast<const void*>(i + 1)));
    }
    const auto doc_map = reconcile_tree_items(
        tree_, doc_keys,
        [](const std::string&) {
            auto* item = new QTreeWidgetItem(
                QStringList{QStringLiteral("未命名图件"), QString()});
            item->setFlags(Qt::ItemIsEnabled | Qt::ItemIsSelectable);
            return item;
        },
        [this](QTreeWidgetItem* item, const std::string& key) {
            const Json* doc = nullptr;
            for (std::size_t i = 0; i < documents_.size(); ++i) {
                if (document_key(
                        documents_[i],
                        reinterpret_cast<const void*>(i + 1)) == key) {
                    doc = &documents_[i];
                    break;
                }
            }
            const std::string name =
                doc != nullptr
                    ? field_value_str(*doc, "name", "未命名图件")
                    : std::string("未命名图件");
            const QString qname = qstr(name.empty() ? "未命名图件" : name);
            if (item->text(0) != qname) {
                item->setText(0, qname);
            }
            item->setData(0, kTagRole, QStringLiteral("doc"));
            item->setData(0, kKeyRole, qstr(key));
        },
        root);
    doc_items_.clear();
    doc_items_.reserve(documents_.size());
    for (const std::string& key : doc_keys) {
        doc_items_.push_back(doc_map.at(key));
    }

    // The active document (or the last one as the fallback) carries the
    // layer subtree; every other document's children are stripped.
    QTreeWidgetItem* populated_item = nullptr;
    const Json* populated_doc = nullptr;
    for (std::size_t i = 0; i < documents_.size(); ++i) {
        if (active_document_ != nullptr &&
            documents_[i] == *active_document_) {
            reconcile_document_children(doc_items_[i], documents_[i]);
            populated_item = doc_items_[i];
            populated_doc = &documents_[i];
        }
    }
    if (active_document_ == nullptr && !doc_items_.empty()) {
        reconcile_document_children(doc_items_.back(), documents_.back());
        populated_item = doc_items_.back();
        populated_doc = &documents_.back();
    }
    for (std::size_t i = 0; i < doc_items_.size(); ++i) {
        if (doc_items_[i] != populated_item) {
            reconcile_tree_items(
                tree_, {},
                [](const std::string&) { return new QTreeWidgetItem(); },
                [](QTreeWidgetItem*, const std::string&) {},
                doc_items_[i]);
        }
    }
    populated_document_ = populated_doc;
    tree_->blockSignals(was_blocked);
    suppress_item_changed_ = false;
}

void MapLayerTree::reconcile_document_children(QTreeWidgetItem* parent,
                                               const Json& document) {
    populated_document_ = &document;
    const Json ref_layers =
        field_value(document, "reference_layers", Json::array());
    const bool has_refs = ref_layers.is_array() && !ref_layers.empty();
    std::vector<std::string> child_keys;
    for (const char* key : kLayerKeys) {
        child_keys.push_back(std::string("layer:") + key);
    }
    if (has_refs) {
        child_keys.push_back(kRefGroupKey);
    }
    const bool first_population = parent->childCount() == 0;
    const auto items = reconcile_tree_items(
        tree_, child_keys,
        [](const std::string& key) {
            if (key == kRefGroupKey) {
                auto* group = new QTreeWidgetItem(
                    QStringList{QStringLiteral("参考图层"), QString()});
                group->setFlags(Qt::ItemIsEnabled);
                return group;
            }
            auto* item = new QTreeWidgetItem(QStringList{QString(), QString()});
            item->setFlags(Qt::ItemIsEnabled | Qt::ItemIsUserCheckable |
                           Qt::ItemIsSelectable);
            return item;
        },
        [this](QTreeWidgetItem* item, const std::string& key) {
            if (key == kRefGroupKey) {
                if (item->text(0) != QStringLiteral("参考图层")) {
                    item->setText(0, QStringLiteral("参考图层"));
                }
                return;
            }
            const std::string layer_key = key.substr(6);  // "layer:" prefix
            const auto& labels = layer_labels();
            const auto it = labels.find(layer_key);
            const QString label =
                qstr(it != labels.end() ? it->second : layer_key);
            if (item->text(0) != label) {
                item->setText(0, label);
            }
            item->setData(0, kTagRole, QStringLiteral("layer"));
            item->setData(0, kKeyRole, qstr(layer_key));
            item->setCheckState(0, layer_is_visible(layer_key)
                                       ? Qt::Checked
                                       : Qt::Unchecked);
            const QString mark =
                layer_is_locked(layer_key) ? QStringLiteral("⊘")
                                           : QString();
            if (item->text(1) != mark) {
                item->setText(1, mark);
            }
        },
        parent);
    layer_items_.clear();
    for (const char* key : kLayerKeys) {
        layer_items_[key] = items.at(std::string("layer:") + key);
    }
    // Only the first population expands (switching the active document or
    // the initial bind); same-key refresh keeps the user's collapse state.
    if (first_population) {
        parent->setExpanded(true);
    }
    if (!has_refs) {
        return;
    }
    QTreeWidgetItem* ref_group = items.at(kRefGroupKey);
    const bool ref_first = ref_group->childCount() == 0;
    std::vector<std::string> ref_keys;
    for (std::size_t i = 0; i < ref_layers.size(); ++i) {
        ref_keys.push_back(reference_layer_key(
            ref_layers.at(i), reinterpret_cast<const void*>(i + 1)));
    }
    reconcile_tree_items(
        tree_, ref_keys,
        [](const std::string&) {
            auto* item = new QTreeWidgetItem(QStringList{QString(), QString()});
            item->setFlags(Qt::ItemIsEnabled | Qt::ItemIsSelectable);
            return item;
        },
        [this, &ref_layers](QTreeWidgetItem* item, const std::string& key) {
            const Json* layer = nullptr;
            for (std::size_t i = 0; i < ref_layers.size(); ++i) {
                if (reference_layer_key(
                        ref_layers.at(i),
                        reinterpret_cast<const void*>(i + 1)) == key) {
                    layer = &ref_layers.at(i);
                    break;
                }
            }
            std::string name =
                layer != nullptr
                    ? field_value_str(*layer, "name", "未命名参考图层")
                    : std::string("未命名参考图层");
            const std::string status =
                layer != nullptr ? field_value_str(*layer, "status", "")
                                 : std::string();
            if (status == "offline") {
                name += " (离线)";
            } else if (status == "failed") {
                name += " (失败)";
            }
            const QString qname = qstr(name);
            if (item->text(0) != qname) {
                item->setText(0, qname);
            }
            item->setData(0, kTagRole, QStringLiteral("reference"));
        },
        ref_group);
    if (ref_first) {
        ref_group->setExpanded(true);
    }
}

void MapLayerTree::on_current_item_changed(QTreeWidgetItem* current) {
    if (current == nullptr) {
        return;
    }
    const QString tag = current->data(0, kTagRole).toString();
    if (tag != QLatin1String("doc")) {
        // Layer / reference rows never change the active document.
        return;
    }
    const std::string key =
        current->data(0, kKeyRole).toString().toStdString();
    const Json* doc = nullptr;
    for (std::size_t i = 0; i < documents_.size(); ++i) {
        if (document_key(documents_[i],
                         reinterpret_cast<const void*>(i + 1)) == key) {
            doc = &documents_[i];
            break;
        }
    }
    if (doc == nullptr) {
        return;
    }
    if (active_document_ == nullptr || *doc != *active_document_) {
        active_document_ = doc;
        rebuild_tree();
        for (QTreeWidgetItem* item : doc_items_) {
            if (item->data(0, kKeyRole).toString().toStdString() == key) {
                tree_->blockSignals(true);
                tree_->setCurrentItem(item);
                item->setExpanded(true);
                tree_->blockSignals(false);
                break;
            }
        }
        emit document_selected(*doc);
    }
}

void MapLayerTree::on_item_changed(QTreeWidgetItem* item, int column) {
    if (suppress_item_changed_ || item == nullptr) {
        return;
    }
    if (item->data(0, kTagRole).toString() != QLatin1String("layer")) {
        return;
    }
    const std::string layer_key =
        item->data(0, kKeyRole).toString().toStdString();
    if (column == 0) {
        const bool visible = item->checkState(0) == Qt::Checked;
        if (layer_visible_[layer_key] != visible) {
            layer_visible_[layer_key] = visible;
            emit layer_visibility_changed(qstr(layer_key), visible);
        }
    }
}

std::vector<std::string> MapLayerTree::document_keys() const {
    std::vector<std::string> keys;
    for (QTreeWidgetItem* item : doc_items_) {
        keys.push_back(item->data(0, kKeyRole).toString().toStdString());
    }
    return keys;
}

std::vector<std::string> MapLayerTree::layer_item_keys() const {
    // Canonical LAYER_KEYS order (Python tuple order) — not the std::map
    // storage order.
    std::vector<std::string> keys;
    keys.reserve(layer_items_.size());
    for (const char* key : kLayerKeys) {
        if (layer_items_.find(key) != layer_items_.end()) {
            keys.emplace_back(key);
        }
    }
    return keys;
}

}  // namespace pwb::ui_map
